docs/custom_promise_types/groups.py

```diff
@@ -0,0 +1,335 @@
+            
+            # check values of attributes in members
+            for users in members.values():
+                for user in users:
+                    match = re.match(self._name_regex, user)
+                    if match == None:
+                        self.log_warning(f"username '{user}' not according to recomendations in system reference manual")
+                    if len(user) > self._name_maxlen:
+                        raise ValidationError(f"username too long: ({len(user)} > {self._name_maxlen})")
+            
+            # make sure a user is not included and excluded
+            if "include" in members and "exclude" in members:
+                include = members["include"]
+                exclude = members["exclude"]
+
+                for included in include:
+                    for excluded in exclude:
+                        if included == excluded:
+                            raise ValidationError(f"user '{included}' both included and excluded from group '{promiser}'")
+
+
+    def evaluate_promise(self, promiser, attributes):
+        failed_repairs = 0
+        repairs = 0
+
+        # get group from /etc/group if present
+        group = None
+        try:
+            group = Group.lookup(promiser)
+        except GroupException as e:
+            self.log_error(f"failed to lookup group '{promiser}': {e}")
+            failed_repairs += 1
+
+        # parse json in attribute members
+        if "members" in attributes:
+            # some black magic which removes back slashes parsing before parsing json
+            members = json.loads(json.loads('"' + attributes["members"] + '"'))
+            attributes["members"] = members
+
+        # set policy to present by default, if not specified
+        if "policy" not in attributes:
+            self.log_verbose(f"policy not specified, defaults to present")
+            attributes["policy"] = "present"
+
+        if attributes["policy"] == "present":
+            self.log_debug(f"group '{promiser}' should exist")
+
+            if group is None:
+                self.log_debug(f"group '{promiser}' does not exist")
+                try:
+                    self.log_info(f"creating group '{promiser}'")
+                    Group.create(promiser)
+                    repairs += 1
+                except GroupException as e:
+                    self.log_error(f"failed to create group '{promiser}': {e}")
+                    failed_repairs += 1
+            else:
+                self.log_debug(f"group '{promiser}' does exist")
+
+            group = Group.lookup(promiser)
+            if "gid" in attributes:
+                gid = attributes["gid"]
+
+                if gid != group.gid:
+                    try:
+                        self.log_info(f"setting GID of group '{group.name}' to '{gid}'")
+                        Group.gid_set(group.name, gid)
+                        repairs += 1
```

Comment:
I'm not quite sure this should be done. This is effectively renaming the existing group not setting GID for a new group. There might be files on the file system with that `gid` and so this could grant permissions to those files to this new group. I think it would be safer to make this an error.

Reply:
Good point ! :+1:

Reply:
It should only create group with GID, if it does not exist? E.g. `groupadd --gid 123  foo`